### ISSUE TITLE
install.packages("torch") results in "as: out of memory allocating 8 bytes after a total of 2080063488 bytes"

### ISSUE THREAD

**Issue by liar666** [mlverse/torch#724]
Hi,

I have a machine with 12 GB of RAM, When I close all apps, I have 74% of it free.

Trying to install torch very quickly fails with memory allocation error (given in title and below):
```
> install.packages(c("torch"), ask=F)
trying URL 'https://cran.biotools.fr/src/contrib/torch_0.6.0.tar.gz'
Content type 'application/x-gzip' length 1515965 bytes (1.4 MB)
==================================================
downloaded 1.4 MB

* installing *source* package ‘torch’ ...
** package ‘torch’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
g++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG  -I'/home/gmuller/.R/x86_64-pc-linux-gnu-library/4.1/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-kbcSEe/r-base-4.1.1=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o

as: out of memory allocating 8 bytes after a total of 2080063488 bytes
/tmp/ccCNR9HF.s: Assembler messages:
/tmp/ccCNR9HF.s: Fatal error: RcppExports.o: memory exhausted
make: *** [/usr/lib/R/etc/Makeconf:177: RcppExports.o] Error 1
ERROR: compilation failed for package ‘torch’
* removing ‘/home/gmuller/.R/x86_64-pc-linux-gnu-library/4.1/torch’
* restoring previous ‘/home/gmuller/.R/x86_64-pc-linux-gnu-library/4.1/torch’

The downloaded source packages are in
	‘/tmp/RtmplGadw2/downloaded_packages’
Warning message:
In install.packages(update[instlib == l, "Package"], l, contriburl = contriburl,  :
  installation of package ‘torch’ had non-zero exit status
Calls: update.packages -> install.packages
```

With installing Python package I also have memory errors (rather disk related: "No space left on device"), but I can solve them using `--no-cache-dir` and `--build`

Is there a similar solution to install RTorch ?

**Comment by dfalbel** [comment]
It [looks like](https://stackoverflow.com/questions/45605806/out-of-memory-allocating-bytes) this error can be caused by either disk or memory. 

Do you have enough disk space to install torch? I don't think you need more than 8GB to compile torch from source.
One alternative is to use pre-built binaries for your linux distribution from: https://packagemanager.rstudio.com/client/#/repos/1/overview

**Comment by hsbadr** [comment]
> No space left on device

What's the output of `df -h /tmp`? It looks like your `/tmp` disk is full. Try to empty it or set `TMPDIR` environment variable to another disk, before starting `R`.

**Comment by mediware** [comment]
I have been trying to install torch on debian Buster unsuccessfully. I did it either as admin or user and still no success. In the end of installation effort, it comes with the following error.
--
/usr/lib/R/lib -lR
Renaming torch lib to torchpkg
"/usr/lib/R/bin/Rscript" "../tools/renamelib.R"
Error: Could not patch RcppExports.R, looked for files in: 
~/../R/RcppExports.R
~/../R/RcppExports.R
Execution halted
make: *** [Makevars:5: rename] Error 1
ERROR: compilation failed for package ‘torch’
* removing ‘/usr/local/lib/R/site-library/torch’

The downloaded source packages are in
	‘/tmp/RtmpZP3Fil/downloaded_packages’
Warning message:
In install.packages("torch") :
  installation of package ‘torch’ had non-zero exit status
--

**Comment by liar666** [comment]
Hi,

Sorry for the late response. Problem still occurs and solution proposed here did not solve it.

> Do you have enough disk space to install torch? I don't think you need more than 8GB to compile torch from source. One alternative is to use pre-built binaries for your linux distribution from: https://packagemanager.rstudio.com/client/#/repos/1/overview

I have 4GB free on `/` and 45GB on `/home`... 
As `/` status does not change for either partition during the installation attempt, I don't think this is the problem... More probably a problem of RAM.

**Comment by liar666** [comment]
> > No space left on device

If you read my message carefully, I was saying that this is **not** the message I have.

> What's the output of `df -h /tmp`? It looks like your `/tmp` disk is full. Try to empty it or set `TMPDIR` environment variable to another disk, before starting `R`.

As I replied previously, I have 4GB+ on `/tmp` and this amount does not change during the installation attempt, so it does not seems to be the problem.
```
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda4       115G  105G  4.3G  97% /
```

**Comment by liar666** [comment]
@mediware You problem is totally different from the one addressed here. If you read your error message carefully it says it's a "patching" error:
```
Error: Could not patch RcppExports.R, looked for files in:
~/tmp/../R/RcppExports.R
~/tmp/R/RcppExports.R
```
Please open another issue.

**Comment by liar666** [comment]
SOLUTION FOUND!

Just in case someone gets the same problem... Here is it:

- PROBLEM

For some probably obscure security reason, there was a low ulimit on several memory-related entries:
```
$ ulimit -a
...
max locked memory           (kbytes, -l) 1519764
max memory size             (kbytes, -m) 2097152
...
```
- SOLUTION
I went `root` (as users can only decrease the values):  
```
# ulimit -l $(($(ulimit -l)*1024))
```
Then I (faked a) "login" as user and ran the R pkg upgrade:
```
# su - <user>
$ R
>  install.packages("torch", ask=F, repos="https://packagemanager.rstudio.com/all/__linux__/focal/latest")
```

Just in case someone gets the same problem...
